datafusion/execution/src/registry.rs

```diff
@@ -45,7 +48,26 @@ pub trait FunctionRegistry {
         not_impl_err!("Registering ScalarUDF")
     }
 
-    // TODO add register_udaf and register_udwf
+    /// Registers a new [`AggregateUDF`], returning any previously registered
+    /// implementation.
+    ///
+    /// Returns an error (the default) if the function can not be registered,
+    /// for example if the registry is read only.
+    fn register_udaf(
```

Comment:
This is the key (backwards compatible) API addition in this PR

Reply:
I'm wondering why this is a backwards compatible API? Does `FunctionRegistry` provide this API before?

Reply:
I think it is backwards compatible because it provides a default implementation (to return a Not Yet Implemented error) 

Thus, any existing implementations of `FunctionRegistry` will continue to compile and work as they did before, without any required code changes